martian-pipestance/src/critical_path.rs

```diff
@@ -0,0 +1,308 @@
+
+    fn new() -> Self {
+        Self::default()
+    }
+
+    fn build(
+        mut self,
+        final_state: &FinalState,
+        perf: &Perf,
+        weight_function: StageWeight,
+    ) -> CriticalPath {
+        for perf_element in perf.0.iter().filter(|p| p.ty == NodeType::Stage) {
+            self.add_node(&perf_element.fqname, weight_function.weight(perf_element));
+        }
+
+        let return_path_map: BTreeMap<_, _> = final_state
+            .completed_stages()
+            .flat_map(|f| {
+                f.return_bindings()
+                    .filter_map(|r| r.value.as_ref().map(collect_all_nested_strings))
+                    .flatten()
+                    .filter_map(|ret_val| {
+                        ret_val
+                            .contains(&format!("/{}/fork", f.name))
+                            .then_some((ret_val, f.fqname.clone()))
+                    })
+            })
+            .collect();
+
+        for stage_state in final_state.completed_stages() {
+            assert!(self.node_id_of_stage_id.contains_key(&stage_state.fqname));
+
+            for binding_info in stage_state.argument_bindings() {
+                match binding_info.mode {
+                    ArgumentMode::Empty => {
+                        if let Some(v) = &binding_info.value {
+                            for parent in collect_all_nested_strings(v)
+                                .into_iter()
+                                .filter_map(|path| return_path_map.get(&path))
+                            {
+                                self.add_edge(parent, &stage_state.fqname);
+                            }
+                        }
+                    }
+                    ArgumentMode::Reference => {
+                        self.add_edge(binding_info.node.as_ref().unwrap(), &stage_state.fqname);
+                    }
+                }
+            }
+        }
+        self._critical_path()
+    }
+
+    fn add_node(&mut self, stage_id: &str, weight: f64) {
+        let node = GraphNode {
+            weight: weight.into(),
+            stage_id: stage_id.to_string(),
+        };
+        assert!(!self.node_id_of_stage_id.contains_key(&node.stage_id));
+        self.node_id_of_stage_id
+            .insert(node.stage_id.to_string(), self.graph.add_node(node));
+    }
+
+    fn add_edge(&mut self, from: &str, to: &str) {
+        self.graph.update_edge(
+            self.node_id_of_stage_id[from],
+            self.node_id_of_stage_id[to],
+            (),
+        );
+    }
+
+    fn add_start_and_end_node(&mut self) {
+        let all_nodes: Vec<_> = self.node_id_of_stage_id.keys().cloned().collect();
+
+        self.add_node(Self::START_NODE, 0.0);
+        self.add_node(Self::END_NODE, 0.0);
+
+        for node in all_nodes {
+            self.add_edge(Self::START_NODE, &node);
+            self.add_edge(&node, Self::END_NODE);
+        }
+    }
+
+    fn _critical_path(mut self) -> CriticalPath {
```

Comment:
Leading underscore method names aren't idiomatic in Rust; even then, this method is called and is already marked private.
```suggestion
    fn critical_path(mut self) -> CriticalPath {
```